File: cpp/09_special_pythagorean_triplet.cpp
#include <iostream>

int special_pythagorean_triplet(){
    // a^2 + b^2 = c^2, a + b + c = 1000
    // c = 1000 - a - b
    // a^2 + b^2 = (1000 - a - b)^2
    
    // let a be the shortest side
    // a is at most 332
    // minimal triplet with a = 333: 333 + 334 + 335 > 1000
    // b is at most 499
    // minimal triplet with b = 500: 1 + 500 + 501 > 1000
    for (int a = 1; a < 333; a++){
        for (int b = a + 1; b < 500; b++){
            int c = 1000 - a - b;
            if (a*a + b*b == c*c) return a*b*c;
        }
    }
    return -1;
}

int main() {
    std::cout << special_pythagorean_triplet() << "\n";
}

File: c#/_09_SpecialPythagoreanTriplet.cs
using System;

public class _09_SpecialPythagoreanTriplet{
    public static int Euler09(){
        // a + b + c = p, a^2 + b^2 = c^2
        // c = p - a - b
        
        // a^2 + b^2 = (p - a - b)^2
        // a^2 + b^2 = p^2 - ap - bp - ap + a^2 + ab - bp + ab + b^2
        // 0 = p^2 - 2ap - 2bp + 2ab
        // 2bp - 2ab = p^2 - 2ap
        // b * (2p - 2a) = p^2 - 2ap
        // b = (p^2 - 2ap) / (2p - 2a)
        
        int p = 1000;
        for (int a = 1; a <= p / 3; a++){
            int b = (p * p - 2 * a * p) / (2 * p - 2 * a);
            int c = p - a - b;
            if (a * a + b * b == c * c){
                return a * b * c;
            }
        }
        return 0;
    }
    
    public static void Main(string[] args){
        Console.WriteLine(Euler09());
    }
}
